File: 2022-2023/sourcecode/lab/lab13/OfflineCachingBasicsJava/app/src/main/java/pl/udu/uwr/pum/offlinecachingbasicsjava/di/AppModule.java
package pl.udu.uwr.pum.offlinecachingbasicsjava.di;

import android.app.Application;

import androidx.room.Room;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import pl.udu.uwr.pum.offlinecachingbasicsjava.data.db.UserDatabase;
import pl.udu.uwr.pum.offlinecachingbasicsjava.data.remote.RandomApi;
import pl.udu.uwr.pum.offlinecachingbasicsjava.data.repository.UserRepository;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
@InstallIn(SingletonComponent.class)
public class AppModule {

    @Provides
    @Singleton
    RandomApi provideRandomApi(){
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build();
        return new Retrofit.Builder()
                .baseUrl("https://random-data-api.com/api/v2/")
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build().create(RandomApi.class);
    }

    @Provides
    @Singleton
    UserDatabase provideUserDatabase(Application app){
        return Room.databaseBuilder(app, UserDatabase.class, "user_database_java4").build();
    }

    @Provides
    @Singleton
    UserRepository provideUserRepository(RandomApi api, UserDatabase db){
        return new UserRepository(db.userDao(), api);
    }
}

File: 2022-2023/sourcecode/lab/lab13/OfflineCachingBasicsKotlin/app/src/main/java/pl/udu/uwr/pum/offlinecachingbasicskotlin/di/AppModule.kt
package pl.udu.uwr.pum.offlinecachingbasicskotlin.di

import android.app.Application
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import pl.udu.uwr.pum.offlinecachingbasicskotlin.data.db.UserDatabase
import pl.udu.uwr.pum.offlinecachingbasicskotlin.data.remote.RandomApi
import pl.udu.uwr.pum.offlinecachingbasicskotlin.data.repository.UserRepository
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Provides
    @Singleton
    fun provideRandomApi(): RandomApi{
        val interceptor = HttpLoggingInterceptor()
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val client = OkHttpClient.Builder()
            .addInterceptor(interceptor)
            .build()
        return Retrofit.Builder()
            .baseUrl("https://random-data-api.com/api/v2/")
            .addConverterFactory(GsonConverterFactory.create())
            .client(client)
            .build().create(RandomApi::class.java)
    }

    @Provides
    @Singleton
    fun provideAppRepository(api: RandomApi, db: UserDatabase) : UserRepository {
        return UserRepository(api, db)
    }

    @Provides
    @Singleton
    fun provideUserDatabase(app: Application): UserDatabase = UserDatabase.getDatabase(app)
}